### PR TITLE
implement array_fill

### PR DESCRIPTION
**Pull request by valkrypton**
implemented array_fill under [#4153](https://github.com/diesel-rs/diesel/issues/4153)

### CODE REVIEW

**Comment by valkrypton** [comment]
not sure what arguments should i pass in auto_type.rs function call

File: diesel/src/pg/expression/helper_types.rs
Comment (valkrypton) [inline diesel/src/pg/expression/helper_types.rs:1]: I fixed it by looking at `array_to_string` definition and changing the definition in helper_type.rs according to, but I am not sure why I have to write a generic argument for an argument with a concrete type

File: diesel/src/pg/expression/helper_types.rs [hunk 1]
```diff
@@ -407,10 +407,10 @@ pub type cardinality<A> = super::functions::cardinality<SqlTypeOf<A>, A>;
...
 #[cfg(feature = "postgres_backend")]
-pub type array_fill<E> = super::functions::array_fill<SqlTypeOf<E>, E, Array<Integer>>;
+pub type array_fill<E, A> = super::functions::array_fill<SqlTypeOf<E>, E, A>;
 
 /// Return type of [`array_fill_with_lower_bound(value,array,array)`](super::functions::array_fill_with_lower_bound())
 #[allow(non_camel_case_types)]
 #[cfg(feature = "postgres_backend")]
-pub type array_fill_with_lower_bound<E> =
-    super::functions::array_fill_with_lower_bound<SqlTypeOf<E>, E, Array<Integer>, Array<Integer>>;
+pub type array_fill_with_lower_bound<E, A1, A2> =
```
Comment (valkrypton) [inline diesel/src/pg/expression/helper_types.rs:415]: Not sure why i have to write generics here for arguments that have concrete types, but this works. An explanation would be appreciated!

File: diesel/src/pg/expression/functions.rs [hunk 2]
```diff
@@ -996,3 +996,83 @@ define_sql_function! {
...
+    /// .get_result::<Vec<Option<i32>>>(connection)?;
+    /// assert_eq!(vec![Some(4),Some(4),Some(4)],array);
+    ///
+    /// let array = diesel::select(array_fill::<Nullable<Integer>,_,_>(None::<i32>,vec![3]))
+    /// .get_result::<Vec<Option<i32>>>(connection)?;
+    /// assert_eq!(vec![None::<i32>,None::<i32>,None::<i32>],array);
+    /// # Ok(())
+    /// # }
+    ///
+    fn array_fill<E:SingleValue>(value:E,dim:Array<Integer>) -> Array<E>;
```
Comment (weiznich) [inline diesel/src/pg/expression/functions.rs:1037]:
```suggestion
    fn array_fill<E:SingleValue>(value: E, dim: Array<Integer>) -> Array<E>;
```

File: diesel/src/pg/expression/helper_types.rs (hunk 1 above)
Comment (weiznich) [reply diesel/src/pg/expression/helper_types.rs:415]: `define_sql_function!` generates a function signature as follows for the `array_fill` function:

```rust
fn array_fill<E: SingleValue, value: T1, dim: T2>(value: T1, dim: T2) -> _
where T1: AsExpression<E>, 
            T2: AsExpression<Array<Integer>>
```

The additional `AsExpression` bounds enable diesel to accept a typed SQL expression (like for example a column) or a rust side value (for example `5`) as argument in this position.

**Pull request merged**